community/unsafe/src/main/java/org/neo4j/unsafe/impl/internal/dragons/FeatureToggles.java

```diff
@@ -0,0 +1,131 @@
+    {
+        return booleanProperty( name( location, name ), defaultValue );
+    }
+
+    public static long getLong( String name, long defaultValue )
+    {
+        return HACK == null ? defaultValue : getLong( HACK.getCallingClass(), name, defaultValue );
+    }
+
+    public static long getLong( Class<?> location, String name, long defaultValue )
+    {
+        return Long.getLong( name( location, name ), defaultValue );
+    }
+
+    public static int getInteger( String name, int defaultValue )
+    {
+        return HACK == null ? defaultValue : getInteger( HACK.getCallingClass(), name, defaultValue );
+    }
+
+    public static int getInteger( Class<?> location, String name, int defaultValue )
+    {
+        return Integer.getInteger( name( location, name ), defaultValue );
+    }
+
+    public static <E extends Enum<E>> E flag( String name, E defaultValue )
+    {
+        return HACK == null ? defaultValue : flag( HACK.getCallingClass(), name, defaultValue );
+    }
+
+    public static <E extends Enum<E>> E flag( Class<?> location, String name, E defaultValue )
+    {
+        return enumProperty( defaultValue.getDeclaringClass(), name( location, name ), defaultValue );
+    }
+
+    // <implementation>
+
+    private static String name( Class<?> location, String name )
+    {
+        return location.getCanonicalName() + "." + name;
+    }
+
+    private static boolean booleanProperty( String flag, boolean defaultValue )
+    {
+        return parseBoolean( System.getProperty( flag ), defaultValue );
+    }
+
+    private static boolean parseBoolean( String value, boolean defaultValue )
+    {
+        return defaultValue ? !"false".equalsIgnoreCase( value ) : "true".equalsIgnoreCase( value );
+    }
+
+    private static <E extends Enum<E>> E enumProperty( Class<E> enumClass, String name, E defaultValue )
+    {
+        try
+        {
+            return Enum.valueOf( enumClass, System.getProperty( name, defaultValue.name() ) );
+        }
+        catch ( IllegalArgumentException e )
+        {
+            return defaultValue;
+        }
+    }
+
+    private static final Hack HACK;
+
+    static
+    {
+        Hack hack;
+        try
+        {
+            hack = AccessController.doPrivileged( new PrivilegedAction<Hack>()
+            {
+                @Override
+                public Hack run()
+                {
+                    return new Hack();
+                }
+            } );
+        }
+        catch ( Exception e )
+        {
+            if ( flag( FeatureToggles.class, "LOG_HACK_FAILURE", false ) )
+            {
+                System.err.println( "Failing to access call stack, default values will be used." );
+                e.printStackTrace( System.err );
+            }
+            hack = null;
+        }
+        HACK = hack;
+    }
+
+    private static class Hack extends SecurityManager
+    {
+        Class<?> getCallingClass()
+        {
+            return getClassContext()[2];
```

Comment:
and as soon as you will rename your class IDE will not gonna be able to help you with that, do not see how documenting that behaviour will help to prevent people mistakes, as soon as somebody will pass flag or property as param or similar, property will not be found and will fall back to default value - which will be unnoticeable till it will brake something on live system.